src/microcalibrate/calibration.py

```diff
@@ -463,3 +487,916 @@ def summary(
+        storage: Optional[str] = None,
+        load_if_exists: Optional[bool] = False,
+        direction: Optional[str] = "minimize",
+        sampler: Optional["optuna.samplers.BaseSampler"] = None,
+        pruner: Optional["optuna.pruners.BasePruner"] = None,
+    ) -> Dict[str, Any]:
+        """
+        Tune hyperparameters for L0 regularization using Optuna.
+
+        This method optimizes l0_lambda, init_mean, and temperature to achieve:
+        1. Low calibration loss
+        2. High percentage of targets within 10% of their true values
+        3. Sparse weights (fewer non-zero weights)
+
+        Args:
+            n_trials: Number of optimization trials to run.
+            objectives_balance: Dictionary to balance the importance of loss, accuracy, and sparsity in the objective function. Default prioritizes being within 10% of targets.
+            epochs_per_trial: Number of epochs per trial. If None, uses self.epochs // 4.
+            n_holdout_sets: Number of different holdout sets to create and evaluate on
+            holdout_fraction: Fraction of targets in each holdout set
+            aggregation: How to combine scores across holdouts ("mean", "median", "worst")
+            timeout: Stop study after this many seconds. None means no timeout.
+            n_jobs: Number of parallel jobs. -1 means using all processors.
+            study_name: Name of the study for storage.
+            storage: Database URL for distributed optimization.
+            load_if_exists: Whether to load existing study.
+            direction: Optimization direction ('minimize' or 'maximize').
+            sampler: Optuna sampler for hyperparameter suggestions.
+            pruner: Optuna pruner for early stopping of trials.
+
+        Returns:
+            Dictionary containing the best hyperparameters found.
+        """
+        # Suppress Optuna's logs during optimization
+        optuna.logging.set_verbosity(optuna.logging.WARNING)
+
+        if epochs_per_trial is None:
+            epochs_per_trial = max(self.epochs // 4, 100)
+
+        holdout_sets = self._create_holdout_sets(
+            n_holdout_sets, holdout_fraction, self.seed
+        )
+
+        logger.info(
+            f"Multi-holdout hyperparameter tuning:\n"
+            f"  - {n_holdout_sets} holdout sets\n"
+            f"  - {len(holdout_sets[0]['indices'])} targets per holdout ({holdout_fraction:.1%})\n"
+            f"  - Aggregation: {aggregation}\n"
+        )
+
+        # Store original state
+        original_state = {
+            "excluded_targets": self.excluded_targets,
+            "targets": self.targets.copy(),
+            "target_names": (
+                self.target_names.copy()
+                if self.target_names is not None
+                else None
+            ),
+        }
+
+        # Initialize list to collect all holdout evaluations
+        all_evaluations = []
+
+        def evaluate_single_holdout(
```

Comment:
Pretty complex function to be defined within another function. Feels like this could be standalone logic.